### ISSUE TITLE
lang: Disallow duplicate accounts by default

### ISSUE THREAD

**Issue by armaniferrante** [project-serum/anchor#217]
Anchor programs should error when duplicate accounts are given to a program. We can have some type of attribute to override this check and allow the duplicate, e.g., `#[account(dup)]` or something else. 

Suggested by @jstarry.

**Comment by jstarry** [comment]
Thanks for writing this up @armaniferrante. The most common legitimate use case of duplicate accounts is when an instruction requires a signed authority address which has already been passed an another account input in the instruction. For example, when a stake account authority is itself.

It would be nice if the anchor framework allowed the signing authority use case to work out of the box but still enforced that other account inputs are unique.

**Comment by armaniferrante** [comment]
Would it be sufficient to simply omit the uniqueness check for all accounts marked with the existing `#[account(signer)]` attribute?

**Comment by jstarry** [comment]
This probably a deserves a separate issue but my point was just that some accounts are *only* used for signing and nothing else and so it's totally safe to allow them to be duplicates of other account inputs as long as the framework doesn't allow them to be read/written to.

I think that the `dup` attribute is still useful too

**Comment by WrRaThY** [comment]
To be honest - I think that there are more pressing things than this. It feels like a nice-to-have instead of a real blocker for v1. What would you achieve with this? Less bugs, potentially, but bugs like this should be picked up fairly quickly with e2e tests (and those are kinda mandatory with blockchain development anyway)

**Comment by jstarry** [comment]
I would disagree, these sorts of bugs can lead to pretty serious vulnerabilities and it's easy to miss them. E2E tests don't cover the things you forget to consider

**Comment by armaniferrante** [comment]
Agreed this is important and should be done as soon as possible.

> as long as the framework doesn't allow them to be read/written to.

One challenge is that the `AccountInfo` likely should always be available for CPI and so one can't stop the account from being read, but I think that's ok. If one is parsing data from `AccountInfo` directly, then one is already expected to do extra validation steps.

**Comment by paul-schaaf** [comment]
I'm not sure this is a feature that anchor should have.


In the programs that I write manually, I don't add duplication checks. I don't do it because if someone passes a duplicate account when it's not supposed to be there, my other checks will detect that. I don't see a scenario in which the duplicate check would find a wrong account but my normal constraints checks wouldn't.


As a debug only or a client feature, this would make more sense to me. It would help me find missing constraints or make me aware when building the program that I might not need to pass in the same account twice in an instruction. But once the program is deployed, as I said above, I don't see its value.

**Comment by jstarry** [comment]
> I don't do it because if someone passes a duplicate account (or any account) that is not the account I expect, my other checks will detect that. [...]  Unless of course, I forgot such a check.

Right, the problem is that developers could easily not realize they need other checks like that. I personally think that out-of-the-box Anchor should eliminate footguns like this but allow more experienced devs (like yourself) to opt of of them for better compute efficiency. Devs that start using Anchor directly might not realize that account inputs could be duplicated at all.

> Which is why as a debug only or a client feature, this would make more sense to me. It would help me find missing constraints or make me aware when building the program that I might not need to pass in the same account twice in an instruction. But once the program is deployed, as I said above, I don't see its value.

This sounds more sophisticated to me. Maybe a test framework could compile constraints that are enforced during debug or testing in the future, but I don't think we're there yet. Am I right in my understanding that you're suggesting that such a tool can prove that all paths of a program instruction make it impossible for two of the account inputs to be dupes?

**Comment by paul-schaaf** [comment]
To me it just doesn't quite feel right to make anchor generate code that in my opinion would fit much better into a test.

In general, if I have a function 
```
fn F(x,y) -> Result<(), Error>{
      if !P(x,y) { return Err(...);}
      ...
      Ok(())
}
```
and I think `P(x,y)` (in an anchor program this predicate is my constraints (without dup)) will always be false if `x==y` , I don't add a line before the `P(x,y)` saying `if x==y {return Err(..);}`  (the dup checks) because I think that P(x,y) will already handle this. But i might not be 100% sure it really does so in my testing folder I had a test `assert!(F(42,42).is_err())`.

So to me, the solution to developers not realizing they need more checks is 1) better and more documentation and 2) writing the proper tests. In fact, anchor could even write these tests for me (or at least add test scaffoldings for duplicate accounts tests so I remember writing them). And I feel that learning to write these tests can be just as helpful to new devs as writing the tests for them in the program.

**Comment by armaniferrante** [comment]
I think Anchor should, by default, favor foot-gun elimination over performance--though obviously try to have both and provide an option for experts to omit checks. Furthermore, I think the checks here boil down to a couple if statements in the average case and so the compute overhead is negligible--though I haven't measured. 

On the two ponts. 1) We can generally use more documentation, particularly around educating developers about security, but I don't think we should punt the problem to docs. The tools should be as safe as possible to use by anyone not reading the docs particularly closely--since that will likely be the majority of developers unfortunately.  2) I don't think telling devs to write tests is a great argument, since that applies to all constraints and to typed languages more generally. In practice, compile time guarantees eliminates a lot of bugs, and I think we should do as much of that as possible.

**Comment by paul-schaaf** [comment]
re 2) I think in this particular case telling devs to write tests is not great but better than writing these checks for them. but if it can be turned off, users can decide for themselves, so let's just do that.

**Comment by paul-schaaf** [comment]
we should also decide what to do with `remaining_accounts`. Add another annotation(where?) that allows `remaining_accounts` to contain duplicates of previous accounts? or just not have the dup check for `remaining_accounts`?

**Comment by armaniferrante** [comment]
> we should also decide what to do with `remaining_accounts`. Add another annotation(where?) that allows `remaining_accounts` to contain duplicates of previous accounts? or just not have the dup check for `remaining_accounts`?

Remaining accounts doesn't provide any validation guarantees and the api is the same as regular solana. So I think it's ok to have no checks there, since if you're using it, you know you're already doing unsafe things (relatively speaking).

**Comment by paul-schaaf** [comment]
@armaniferrante you can assign me

**Comment by armaniferrante** [comment]
One way of doing this is to add the uniqueness checks anytime you have accounts of the same type (rather than doing it for all accounts).

**Comment by sushi-shi** [comment]
@armaniferrante yet different types might point to the same account addresses, `Signer` specifically; which might or might not be a problem depending on your specific use case. If done this way it can add additional gotchas, and might be a pain when new types of accounts are added.

**Comment by paul-schaaf** [comment]
@sushi-shi Yea armani and I discussed this offline and agree.

 So my current local implementation will fail only if there are multiple mutable or a mutible and an immutable account (similar to how rust references work), even if they are different types.

**Comment by sushi-shi** [comment]
@paul-schaaf it still sounds like a lot of logic where one shouldn't be. What is the point of having it work like that?

**Comment by paul-schaaf** [comment]
> So my current local implementation will fail only if there are multiple mutable or a mutible and an immutable account (similar to how rust references work), even if they are different types.

@sushi-shi If you have two mutable AccountInfos (or even just one mutable and any number of immutables) pointing to the same account, you might accidently overwrite the account so the idea is that you need to be explicit if you really want the same AccountInfo as mutable twice by using the `dup` constraint.

**Comment by sushi-shi** [comment]
@paul-schaaf my general concern was when multiple accounts can be passed when you don't expect them to be equal (which is almost always true). So maybe it would be the right thing to have immutable accounts to be also checked for this?

**Comment by paul-schaaf** [comment]
@sushi-shi some reasons why we might not want checks for all accounts:

- if we implement them for all accounts, the number of if checks is `(n^2)/2` which will be quite a lot of if checks. if compute units were no issue this wouldnt matter as much but they do. That said, we havent measured this yet so if 50 more if checks dont make a difference, maybe we could do them for all accounts.
- if you have duplicate but only immutable accounts pointing to the same data, your reads will always be consistent, no matter from which account you read the data, because youre not writing to it and you cannot accidently overwrite data from a previous mutation by the other account in the instruction. So intuitively immutable duplicate accounts feel less of a danger than having duplicate mutable accounts.
- the api would have to be more complicated. The mutable duplicate checks are imo more important than the immutable ones so it'd be necessary to allow people to turn off the checks for mutable and immutable duplicate checks separately.

For me the first point is the most important. If it doesn't make a dent in the compute budget at all i'd be down to just have checks for all accounts.

**Comment by paul-schaaf** [comment]
@jstarry @armaniferrante 

interesting issue that came up while implementing this. Some background context first. The way I'm implementing this rn is that the dup checks will also works across `CompositeField`. The parent struct will check that its accounts have no dups with all the accounts in child struct and its child structs. The parent will not check that the accounts in its children have no dups with each other because each child struct has their own generate function taking care of that

The question is what to do when there is an intended dup account. It's easy when the dup constraint (which allows a dup to occur) is in the parent because the parent has access to the child struct. So no problem there. My current api for this is      
   `#[account(dup = child.target)]` (just like the other constraints).

but what if we want a dup constraint is in the child? Should we
- simply not allow this like we do with all other constraints (only the parent can access child accounts)
- add some construction with which children can access their parents (this could then also be used for other constraints)
e.g. `#[account(dup = parent.parent.acc_x)]` on an account in the child. Then the parent checks in their generate function whether they really are the 2nd parent above and have `acc_x` and if so, does not add a no_dup check for that account pair.

Intuitively, I'd say the first option is better. Keeps implementation and API simpler. If people really want to remove the dup checks for this edge case they can inline the child struct. I'd go with the first option and see how users like it and then reevaluate.

**Comment by jstarry** [comment]
I'll leave the decision to @armaniferrante but my perspective is that simple and safe is a great starting point. Adding more complex knobs can come later when devs need them.

**Comment by paul-schaaf** [comment]
@armaniferrante actually, I'm not sure that no_dup checks can be implemented across composite fields at all. no_dup checks require that the generate function of the parent knows the fields of the child but it doesnt cause the generate functions are separate for each `#[derive(accounts)]` so the parent only knows the Ident of the child.

no_dup checks across composite fields could be possible by changing the declaration of instruction accounts structs so that they are all parsed inside one macro call but that may be sth for a different issue.

(dup checks are fine across composite fields because you specify the exact account in the constraint so the parent knows the Ident path to the account in the child)

**Comment by Henry-E** [comment]
A high level summary of this issue and what might solve it
- duplicate checks of program owned accounts marked as mutable. To avoid this issue of [modifying data multiple times](https://learn.blueshift.gg/en/courses/program-security/duplicate-mutable-accounts).
- composite accounts might not be supported or difficult to support
- remaining accounts would probably be ignored

There is a weird edge case where since anchor can't detect when program accounts not marked as mutable are treated as mut, someone could pass in a duplicate mutable account in all their tests and modify it without triggering the duplicate check.

**Comment by robre** [comment]
Here are my 2cents:
- duplicate account bugs aren't that common with anchor. Most duplicate accounts are prevented by writing normal constraints. Most duplicate account issues I've seen are self-transfer issues.
- adding a hidden constraint that no duplicate accounts are allowed by default introduces a new DoS vulnerability vector. 

For example:
Let's say a lending protocol lets you define a referrer as just an `UncheckedAccount` when you sign up. Your referrer should then be passed whenever you borrow. If I then sign up with the referrer being an account that the lending protocol also requires for borrows, for example the main lending config or pool, this user can't use the borrow function because they'd need to pass a duplicate account.

**Comment by chen-robert** [comment]
Superceded by https://github.com/solana-foundation/anchor/issues/3825
